File: Src/Icm.Core.Tests/Collections/SortedCollection/StructKeyStructValue/BaseSortedCollectionTest.vb
Imports Icm.Collections.Generic.StructKeyStructValue

<TestFixture>
Public Class StructKeyStructValue_BaseSortedCollectionTest

    Private Shared Function GetNewCollection() As BaseSortedCollection(Of Long, Char)
        Return New MemorySortedCollection(Of Long, Char)(New LongTotalOrder)
    End Function

    <Test>
    Public Sub GetFreeKey_WithEmptyCollection_ReturnsSameKey()
        Dim coll = GetNewCollection()
        Dim expected As Long = 13

        Dim actual = coll.GetFreeKey(13)

        Assert.That(actual.HasValue)
        Assert.AreEqual(expected, actual)
    End Sub

    <Test>
    Public Sub GetFreeKey_IfKeyNotPresent_ReturnsSameKey()
        Dim coll = GetNewCollection()
        Dim expected As Long = 13

        ' Act
        coll.Add(12, "a"c)
        coll.Add(14, "b"c)
        Dim actual = coll.GetFreeKey(13)

        Assert.That(actual.HasValue)
        Assert.AreEqual(expected, actual)
    End Sub

    <Test>
    Public Sub GetFreeKey_IfKeyPresent_ReturnsNextAvailableKey()
        Dim coll = GetNewCollection()
        Dim expected As Long = 15

        ' Act
        coll.Add(12, "a"c)
        coll.Add(13, "b"c)
        coll.Add(14, "c"c)
        Dim actual = coll.GetFreeKey(13)

        ' Assert
        Assert.That(actual.HasValue)
        Assert.AreEqual(expected, actual)
    End Sub

    <Test>
    Public Sub GetFreeKey_IfRunsOutOfValues_ReturnsNothing()
        Dim coll = GetNewCollection()
        Dim expected As Long? = Nothing

        ' Act
        coll.Add(Long.MaxValue, "a"c)
        Dim actual = coll.GetFreeKey(Long.MaxValue)

        Assert.AreEqual(expected, actual)
    End Sub

    <Test>
    Public Sub ToString_EmptyCollection()
        Dim coll = GetNewCollection()
        Dim expected As String =
            "NC 12 ---" & vbCrLf &
            "NC 14 ---" & vbCrLf

        ' Act
        Dim actual = coll.ToString(12, 14)

        ' Assert
        Assert.AreEqual(expected, actual)
    End Sub

    <Test()>
    Public Sub ToString_OneElementAtStart_AppearsAtStart()
        Dim coll = GetNewCollection()
        Dim expected As String =
            "-> 12 a" & vbCrLf &
            "NC 14 ---" & vbCrLf

        ' Act
        coll.Add(12, "a"c)
        Dim actual = coll.ToString(12, 14)

        Assert.AreEqual(expected, actual)
    End Sub

    <Test()>
    Public Sub ToString_OneElementInTheMiddle_AppearsInTheMiddle()
        Dim coll = GetNewCollection()
        Dim expected As String =
            "NC 12 ---" & vbCrLf &
            "-> 13 a" & vbCrLf &
            "NC 14 ---" & vbCrLf

        ' Act
        coll.Add(13, "a"c)
        Dim actual = coll.ToString(12, 14)

        Assert.AreEqual(expected, actual)
    End Sub


    <Test()>
    Public Sub ToString_OneElementInTheEnd_AppearsInTheEnd()
        Dim coll = GetNewCollection()
        Dim expected As String =
            "NC 12 ---" & vbCrLf &
            "-> 14 a" & vbCrLf

        ' Act
        coll.Add(14, "a"c)
        Dim actual = coll.ToString(12, 14)

        Assert.AreEqual(expected, actual)
    End Sub
End Class

File: Src/Icm.Core.Tests/Collections/SortedCollection/StructKeyStructValue/BaseSortedCollectionTest.cs

using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
using Icm;
using Icm.Collections.Generic.StructKeyStructValue;
using NUnit.Framework;

[TestFixture()]
public class StructKeyStructValue_BaseSortedCollectionTest
{

	private static BaseSortedCollection<long, char> GetNewCollection()
	{
		return new MemorySortedCollection<long, char>(new LongTotalOrder());
	}

	[Test()]
	public void GetFreeKey_WithEmptyCollection_ReturnsSameKey()
	{
		dynamic coll = GetNewCollection();
		long expected = 13;

		dynamic actual = coll.GetFreeKey(13);

		Assert.That(actual.HasValue);
		Assert.AreEqual(expected, actual);
	}

	[Test()]
	public void GetFreeKey_IfKeyNotPresent_ReturnsSameKey()
	{
		dynamic coll = GetNewCollection();
		long expected = 13;

		// Act
		coll.Add(12, 'a');
		coll.Add(14, 'b');
		dynamic actual = coll.GetFreeKey(13);

		Assert.That(actual.HasValue);
		Assert.AreEqual(expected, actual);
	}

	[Test()]
	public void GetFreeKey_IfKeyPresent_ReturnsNextAvailableKey()
	{
		dynamic coll = GetNewCollection();
		long expected = 15;

		// Act
		coll.Add(12, 'a');
		coll.Add(13, 'b');
		coll.Add(14, 'c');
		dynamic actual = coll.GetFreeKey(13);

		// Assert
		Assert.That(actual.HasValue);
		Assert.AreEqual(expected, actual);
	}

	[Test()]
	public void GetFreeKey_IfRunsOutOfValues_ReturnsNothing()
	{
		dynamic coll = GetNewCollection();
		long? expected = null;

		// Act
		coll.Add(long.MaxValue, 'a');
		dynamic actual = coll.GetFreeKey(long.MaxValue);

		Assert.AreEqual(expected, actual);
	}

	[Test()]
	public void ToString_EmptyCollection()
	{
		dynamic coll = GetNewCollection();
		string expected = "NC 12 ---" + Constants.vbCrLf + "NC 14 ---" + Constants.vbCrLf;

		// Act
		dynamic actual = coll.ToString(12, 14);

		// Assert
		Assert.AreEqual(expected, actual);
	}

	[Test()]
	public void ToString_OneElementAtStart_AppearsAtStart()
	{
		dynamic coll = GetNewCollection();
		string expected = "-> 12 a" + Constants.vbCrLf + "NC 14 ---" + Constants.vbCrLf;

		// Act
		coll.Add(12, 'a');
		dynamic actual = coll.ToString(12, 14);

		Assert.AreEqual(expected, actual);
	}

	[Test()]
	public void ToString_OneElementInTheMiddle_AppearsInTheMiddle()
	{
		dynamic coll = GetNewCollection();
		string expected = "NC 12 ---" + Constants.vbCrLf + "-> 13 a" + Constants.vbCrLf + "NC 14 ---" + Constants.vbCrLf;

		// Act
		coll.Add(13, 'a');
		dynamic actual = coll.ToString(12, 14);

		Assert.AreEqual(expected, actual);
	}


	[Test()]
	public void ToString_OneElementInTheEnd_AppearsInTheEnd()
	{
		dynamic coll = GetNewCollection();
		string expected = "NC 12 ---" + Constants.vbCrLf + "-> 14 a" + Constants.vbCrLf;

		// Act
		coll.Add(14, 'a');
		dynamic actual = coll.ToString(12, 14);

		Assert.AreEqual(expected, actual);
	}
}

//=======================================================
//Service provided by Telerik (www.telerik.com)
//Conversion powered by NRefactory.
//Twitter: @telerik
//Facebook: facebook.com/telerik
//=======================================================
